### ISSUE TITLE
Prevent the Editor from Loosing Focus on Blur

### ISSUE THREAD

**Issue by smithjustinb** [facebook/draft-js#726]
As described in #696, when focus is lost in the editor the current selection is also lost.  I can see that this is done intentionally in editOnBlur.js to address a perceived bug in webkit browsers where the browser does not do this automatically.  I would like to request a means to override this functionality and/or a clever suggestion to get around this behavior.

While setting event.preventDefault in an onMouseDown handler works for buttons and the like, select boxes and text fields need focus to function.  My editor requires a list of fonts that users can choose from, and fields such as font size in which an arbitrary value may be entered.  Currently I am saving the selection before focus is lost, and restoring it after a setting is changed.  The user experience is not ideal. At best, the user cannot see the selection to which their setting change will be applied while they are in the process of changing it.  At worst, the user will erroneously think their selection has been lost.  

I believe I can get around the select box issue by rolling my own out of divs, though I'd rather invest the time in other development.  For the text input (font size) case, I'm not coming up with a workable solution.

**Comment by smithjustinb** [comment]
Realized that webkit does deselect for the text input case anyway, so while it would still be nice to be able to set Draft's behavior with a prop for the button and select box cases, I will need a workaround regardless.  Using css to apply a background as a pseudo-highlight for the selection before it disappears is mostly working.

**Comment by ericbiewener** [comment]
I'm doing the same -- create an entity on the selected text with a class that gets styled to match the highlight bg color. I then remove that entity when the editor regains focus again.

**Comment by cbothner** [comment]
Can you explain how you're styling the selection when the editor is blurred? Is that a property of a `customStyleMap`?

**Comment by smithjustinb** [comment]
I'm doing it with a customStyleMap entry that sets the background color.  The one downside that haven't found a workaround for is that, if you have mixed font sizes/line-heights, the "pseudo-selection" highlight will also vary in height, where as actual OS highlighting will remain consistent throughout.

**Comment by cbothner** [comment]
Is there a particular `customStyleMap` key (an analog to `'BOLD'`) that the editor already knows about, or do you have to hack the editor to apply a certain key's style?

**Comment by smithjustinb** [comment]
Nope. Totally custom.  I apply the key as an inline style for the current selection when the editor blurs, and remove it when focus returns.

**Comment by toblix** [comment]
Is this the issue that causes the editor to lose its inline style override when blurring? For example, in the [example editor](http://draftjs.org/), if you select bold, italic, etc. and then click outside the editor, you effectively lose these choices. Stranger still: if you click outside the editor to blur it, then select bold and _then_ click inside the editor, it also clears the style override. (I'm on Chrome, by the way)

The [example editor](http://draftjs.org/) sort of works because the inline styles are pure toggles that prevent focus loss by `e.preventDefault()`-ing the `onMouseDown` events, but when I try to change the inline style in a way that causes the editor to lose focus (like, say, a fancy color picker) the editor loses track of the current inline style and I get sad.

Is it wrong to expect the current inline style override to remain untouched in these cases?

If this is the wrong issue, or if it should be filed as a separate issue, please let me know!

**Comment by smithjustinb** [comment]
I think that's separate, though related (admittedly it's been a while since I looked closely at this). It would be nice to have a means of overriding that behavior as well.  My workaround was to keep track of the last seen inline style in my store (either the style of the last selection, or the last inline override).  When the selection changes, I get the current inline style and if one isn't set I re-apply the last seen.

**Comment by blackChef** [comment]
If you look at the [RichUtil.toggleInlineStyle source code](https://github.com/facebook/draft-js/blob/643c388de646945af3aecb4a362016a8103ade11/src/model/modifier/RichTextEditorUtil.js#L317). It relies on `selectionState` to apply inline style. Because there's no selection after editor blured, it can't apply inline style. 
So I save the `editorState` when `editor.onBlur` triggered, and then pass this `beforeBlurState` to my style modifiers. Now I can update inline style even when editor is blured. But then if I manually focus editor through `editor.focus` or `EditorState.forceSelection`, style lost. Any one can help?

**Comment by lcc19941214** [comment]
> If you look at the [RichUtil.toggleInlineStyle source code](https://github.com/facebook/draft-js/blob/643c388de646945af3aecb4a362016a8103ade11/src/model/modifier/RichTextEditorUtil.js#L317)...

@blackChef 
From the source code, I RichUtils use `EditorState.setInlineStyleOverride`. As the docs says, this method can **returns a new EditorState object with the specified DraftInlineStyle applied as the set of inline styles to be applied to the next inserted characters**.
So manually use `setInlineStyleOverride` may solve your problem.

**Comment by Omhet** [comment]
If you have an editor and all toolbar stuff (buttons, selects, etc) in one parent div, you can do this:
```
<div onFocus={this.focus}>
  <ToolBar ... />
  <Editor ref={this.editorRef} ... />
</div>

focus = () => {
  const editor = this.editorRef.current;
  if (editor) {
    editor.focus();
  }
};
```

**Comment by Davidkle** [comment]
I applied inline styles to the selection during `onBlur` and remove it during `onFocus`. I found that there's a disconnect between `onChange` and the other two event handlers. You can't really use `setState` in `onBlur` and `onFocus` and expect `onChange` to pick it up. You have to use ivars. 

Here's roughly the pseudo code.

```
export enum InlineStyleType {
  SELECTION_TEXT_COLOR = 'SELECTION_TEXT_COLOR',
}

class Container extends React.Component<Props, State> {

  _hasFocus = false;
  _hasSelectionOnBlur = false;
  _onBlurEditorState?: EditorState;

  state: State = {
    editorState: EditorState.createEmpty(),
    styleMap: {
      [InlineStyleType.SELECTION_TEXT_COLOR]: {
        backgroundColor: '#CDE2FF',
      },
    },
  };

  onChange = (newEditorState: EditorState) => {
    const currentContent = newEditorState.getCurrentContent();

    let updatedEditorState: EditorState;

    if (this._hasSelectionOnBlur && this._hasFocus && this._onBlurEditorState) {
      const noSelectionContentState = Modifier.removeInlineStyle(
        currentContent,
        this._onBlurEditorState.getSelection(),
        InlineStyleType.SELECTION_TEXT_COLOR
      );
      updatedEditorState = EditorState.push(newEditorState, noSelectionContentState, 'change-inline-style');

      this._hasFocus = false;
      this._hasSelectionOnBlur = false;
      this._onBlurEditorState = undefined;
    } else if (this._hasSelectionOnBlur) {
      updatedEditorState = RichUtils.toggleInlineStyle(newEditorState, InlineStyleType.SELECTION_TEXT_COLOR);
    } else {
      updatedEditorState = newEditorState;
    }

    this.setState(() => {
      return { editorState: updatedEditorState }
    });
  };

  onEditorBlur = (e: SyntheticEvent) => {
    const { editorState } = this.state;
    if (editorState.getSelection().isCollapsed() === false) {
      this._hasSelectionOnBlur = true;
      this._onBlurEditorState = editorState;
    }

    this._hasFocus = false;
  };

  onEditorFocus = (e: SyntheticEvent) => {
    this._hasFocus = true;
  };

  render() {
    const { editorState, styleMap } = this.state;
    return (
      <DraftJsEditor
        customStyleMap={styleMap}
        editorState={editorState}
        onBlur={this.onEditorBlur}
        onFocus={this.onEditorFocus}
        onChange={this.onChange}
      />
    );
  }
}
```

**Comment by horaciosystem** [comment]
I was able to implement that using most of the code @Davidkle have shared. The only big difference is that I moved all the logic related to applying the inline styles from inside onChange to a useEffect hook, because I was having a stale state inside the onChange function.
